drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,410 @@
+                // ones that we care about
+                let enable: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_EN_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let reset: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_RESET_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let lpmode: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_CTRL_LPMODE_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let present: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_STATUS_PRESENT_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+                let irq: U16<BigEndian> = fpga
+                    .read(Addr::QSFP_STATUS_IRQ_H)
+                    .map_err(|_e| Error::ReadFailed)?;
+
+                // Write one bitfield per active port in the ModuleId
+                let mut count = 0;
+                for mask in modules.ports.to_indices().map(|i| 1 << i) {
+                    let mut status = Status::empty();
+                    if (enable.get() & mask) != 0 {
+                        status |= Status::ENABLED;
+                    }
+                    if (reset.get() & mask) != 0 {
+                        status |= Status::RESET;
+                    }
+                    if (lpmode.get() & mask) != 0 {
+                        status |= Status::LOW_POWER_MODE;
+                    }
+                    if (present.get() & mask) != 0 {
+                        status |= Status::PRESENT;
+                    }
+                    if (irq.get() & mask) != 0 {
+                        status |= Status::INTERRUPT;
+                    }
+                    // Convert from Status -> u8 and write to the output buffer
+                    out[count] = status.bits();
+                    count += 1;
+                }
+                Ok((HostResponse::Status, count))
+            }
+            HostRequest::Read(mem) => {
+                let out_size = mem.len() as u32 * modules.ports.0.count_ones();
+                if out_size as usize > transceiver_messages::MAX_MESSAGE_SIZE {
+                    return Err(Error::RequestTooLarge);
+                }
+                self.read(mem, modules, out)?;
+                Ok((HostResponse::Read(mem), out_size as usize))
+            }
+            HostRequest::Write(mem) => {
+                let data_size = mem.len() as u32 * modules.ports.0.count_ones();
+                // TODO: check equality here and return a different error?
+                if data_size as usize > data.len() {
+                    return Err(Error::RequestTooLarge);
+                }
+                self.write(mem, modules, data)?;
+                Ok((HostResponse::Write(mem), 0))
+            }
+            HostRequest::SetPowerMode(mode) => {
+                let mask = get_mask(modules)?;
+                // TODO: do we need delays in between any of these operations?
```

Comment:
Given that the FPGA will soon manage the details of these transitions (per RFD 244), I don't think so.

Reply:
Added a comment to that effect